ssh/ssh.go

```diff
@@ -47,14 +56,59 @@ func parseShellCommand(command string, existing string) (ssh string, cmd string,
 	return
 }
 
+func findVariant(variant string) (bool, sshVariant) {
+	switch variant {
+	case "ssh", "simple", "putty", "tortoiseplink":
+		return false, sshVariant(variant)
+	case "plink":
+		return false, variantPutty
+	case "auto":
+		return true, ""
+	default:
+		return false, variantSSH
+	}
+}
+
+func autodetectVariant(osEnv config.Environment, gitEnv config.Environment, basessh string) sshVariant {
+	if basessh != defaultSSHCmd {
+		// Strip extension for easier comparison
+		if ext := filepath.Ext(basessh); len(ext) > 0 {
+			basessh = basessh[:len(basessh)-len(ext)]
+		}
+		if strings.EqualFold(basessh, "plink") {
+			return variantPutty
+		}
+		if strings.EqualFold(basessh, "tortoiseplink") {
+			return variantTortoise
+		}
+	}
+	return "ssh"
+}
+
+func getVariant(osEnv config.Environment, gitEnv config.Environment, basessh string) sshVariant {
+	variant, ok := osEnv.Get("GIT_SSH_VARIANT")
+	autodetect, val := findVariant(variant)
+	if ok {
+		if autodetect {
+			return autodetectVariant(osEnv, gitEnv, basessh)
+		}
+		return val
+	}
+	variant, ok = gitEnv.Get("ssh.variant")
+	if autodetect, val := findVariant(variant); ok {
+		if autodetect {
+			return autodetectVariant(osEnv, gitEnv, basessh)
+		}
+		return val
+	}
+	return autodetectVariant(osEnv, gitEnv, basessh)
```

Comment:
There was some tricky corner case I was trying to handle here, but I think I encapsulated that in the tests, so I think your change is a strict improvement here since it's much simpler.